### PR TITLE
feat: Add AST serialization and exporting support

### PR DESCRIPTION
**Pull request by YaCodesDevelopment**
Added `--save_ast` flag. When passed, this should also output the `ast.json` file while compiling. The file includes an array of ASTs for all the files compiled.
This allows the use of that `ast.json` for further analysis of code in third-party libraries, unlocking the ability to create linters and other helper tools.

### CODE REVIEW

File: parser/src/lib.rs [hunk 1]
```diff
@@ -109,6 +116,11 @@ pub fn run_parser(
         }
     }
 
+    if save_ast {
+        let ast_list= serde_json::to_string(&ast_list).unwrap();
```
Comment (Arvolear) [inline parser/src/lib.rs]:
```suggestion
        let ast_list = serde_json::to_string(&ast_list).unwrap();
```

File: circom/src/input_user.rs [hunk 2]
```diff
@@ -344,6 +351,41 @@ mod input_processing {
...
+                            PathBuf::from(path.to_str().unwrap())
+                        }
+                    }
+                    else {
+                        PathBuf::from("ast.json")
+                    }
+               }
+               
+            false => PathBuf::from("ast.json"),
+        }
```
Comment (Arvolear) [inline circom/src/input_user.rs]:
```suggestion
        match matches.is_present("save_ast"){
            true => 
               {
                   let path = matches.value_of("save_ast").unwrap();
                   if path != "" {
                        let path = Path::new(path);
                        if let Some(parent) = path.parent() {
                            std::fs::create_dir_all(parent).unwrap();
                        }
                        
                        if path.is_dir() {
                            PathBuf::from(path.join("ast.json").to_str().unwrap())
                        } else {
                            PathBuf::from(path.to_str().unwrap())
                        }
                    } else {
                        PathBuf::from("ast.json")
                    }
               }
               
            false => PathBuf::from("ast.json"),
        }
```

**Review by Arvolear — APPROVED** [review]
LGTM! (let's wait for the second approval)

File: circom/src/input_user.rs [hunk 3]
```diff
@@ -344,6 +351,41 @@ mod input_processing {
         }
     }
 
+    pub fn get_save_ast(matches: &ArgMatches) -> bool {
+        matches.occurrences_of("save_ast") > 0
+    }
+
+    pub fn get_ast_path(matches: &ArgMatches) -> PathBuf {
```
Comment (velykodnyi) [inline circom/src/input_user.rs]: Return Result and handle errors correctly(like you have in get_prime)

File: parser/src/lib.rs [hunk 4]
```diff
@@ -109,6 +117,11 @@ pub fn run_parser(
         }
     }
 
+    if save_ast {
+        let ast_list = serde_json::to_string(&ast_list).unwrap();
```
Comment (velykodnyi) [inline parser/src/lib.rs]: Panics again

**Review by velykodnyi — CHANGES_REQUESTED** [review]

**Review by velykodnyi — APPROVED** [review]
LGTM

**Pull request merged**